enhancements/20230421-adjust-the-scale-of-resource-quota-during-migration.md

```diff
@@ -0,0 +1,219 @@
+
+`[harvesterhci.io/migratingVMs](http://harvesterhci.io/migratingVMs)` it's a type of `map[string]ResourceList`, where the key is the VM name, and the value is the Kubernetes core resource list type.(`[ResourceList` struct](https://github.com/kubernetes/api/blob/8360d82aecbc72aa039281a394ebed2eaf0c0ccc/core/v1/types.go#L5548-L5549)).
+
+## Design
+
+### Implementation Overview
+
+**VM Migration**
+
+The Migration Controller watches the status of the migration process. When the migration status is Pending, it automatically scales up the ResourceQuota to meet the resource requirements of the VM migration. After the migration is complete, it automatically scales down the quota.
+
+In addition, validators are added to the VM and Restore CRs to ensure that there are sufficient resources to create, start, or restore a VM.
+
+When a migration is created, the Migration Controller adds an annotation to the ResourceQuota in its namespace, as follows:
+
+```yaml
+apiVersion: v1
+kind: ResourceQuota
+metadata:
+  annotations:
+    harvesterhci.io/migratingVMs: '{"vm-01":{"limits.cpu":"1","limits.memory":"1267028Ki"}}'
+  labels:
+    cattle.io/creator: norman
+    resourcequota.management.cattle.io/default-resource-quota: "true"
+  name: default-x7fpr
+  namespace: rs
+spec:
+  hard:
+    limits.cpu: "4"
+    limits.memory: 4952Mi
+status:
+  hard:
+    limits.cpu: "4"
+    limits.memory: 4952Mi
+  used:
+    limits.cpu: "4"
+    limits.memory: 5068112Ki
+```
+
+The key is `vm_name` with a value of the corresponding VM's Pod resources limits. When Migration is in Pending, Harvester should automatically scale up the ResourceQuota to meet the resource requirements for the VM migration. Once the VM migration is complete, Harvester should automatically scale down and delete the key.
+
+Due to the unreliability of the Webhook, the VM Controller will perform a second verification for VMs. If the resources are insufficient, the VM's RunStrategy will be changed to Halted, and an event will be sent indicating the lack of resources.
+
+**Upgrade**
+
+The implementation is similar to the above. If a large number of VM migrations occur in a node, causing the ResourceQuota to scale up and Migration to be in Pending, the VMs that are already started may race for resources. To avoid this situation, the actual available resources for the VMs are calculated using the formula: actual limit - (used - migrations), and the result is compared to see if there are enough resources for the VM to start.
```

Comment:
```suggestion
The implementation is similar to the above. If many VM migrations occur in a node, causing the `ResourceQuota` to scale up and the VM Migration is **Pending**, the VMs that already started may race for resources. To avoid this situation, calculate the actual available resources for the VMs using this formula: **actual limit - (used - migrations)**. Compare this result with the resource quota to see if there are enough resources for the VM to start.
```